MM-1356 Changing team URLs from team.domain.com to domain.com/team

crspeller opened the pull request:
### Main Changes
This pull request changes how URLs in Mattermost operate. 

Team URLs are now of the form domain.com/team/*

The Domain column in the Team table has been renamed to URLId to better reflect its purpose. A column has also been added to the sessions table. Therefore when transitioning *all sessions should be expired*. (or else you will be redirected incorrectly) The transition has been tested, but backups are suggested. 

A response header of X-Team-URL has been added to make it easier and more consistent for the client to get the url in some instances. Not all instances use this new method. This could be a future improvement. 

The signup flow has also been modified:
- When accessing domain.com for the first time, your will be presented with a signup a team page. Future visits with a valid session with automatically redirect you to your team. 
- To sign up additional teams an option has been added to the settings menu where additional teams eventually show up. 
- @asaadmahmoodspin I had to change the width of the signup flow pages because there was not enough horizontal room on the TeamURL page. You might want to take a look at that. 

Code relating to sharding and the subdomain way of organizing teams has mostly been removed. I moved that code to sharding.go, but we can remove it if we don't want it. (it would still be in the git history)

### Additional Changes
- This also fixes MM-1576, clicking on the profile picture now opens the settings menu. 
- Signup flow items now autofocus.
- Signup flow now accepts enter to move to next screen. :tada: 

### Testing
There will be bugs. I would be especially careful of the signup flow. Check for any buttons that redirect improperly. If there is anywhere you click and don't end up where you expected (or end up in town square) report a bug.

coreyhulen commented on api/context.go:
```diff
@@ -64,6 +69,7 @@ type handler struct {
 	requireSystemAdmin bool
 	isApi              bool
 	isUserActivity     bool
+	isTeamIndependent  bool
```
Not sure why this is needed?

crspeller replied:
Team independent handlers will have their teamURL set without a team in it.

coreyhulen commented on api/context.go:
```diff
@@ -135,6 +140,20 @@ func (h handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+		c.Path = "/" + strings.Join(splitURL[2:], "/")
+	}
+
+	w.Header().Set(model.HEADER_TEAM_URL, c.TeamUrl)
```
why set in the header?  didn't this come from the request?

crspeller replied:
This was originally added when I was keeping consistancy between the two url modes. I kept it in case we ever do want to support both. Also fixing the client to always use this header would make things cleaner instead of parsing window.location.

coreyhulen commented on model/team.go:
```diff
@@ -19,7 +19,7 @@ type Team struct {
 	DeleteAt       int64  `json:"delete_at"`
 	Name           string `json:"name"`
-	Domain         string `json:"domain"`
+	URLId          string `json:"urlid"`
```
I don't like the name.  Seems weird to have urlId (We're also inconsistent with how we name Url/URL).  This is really team name and the current team name is the DisplayName.  I think we should bite the bullet and rename so it matches channel.

crspeller replied:
The name comes from the fact that it is not the team name. It is a separate field that might have nothing to do with the team name. It is user configurable. So someone could call their team "My Company" and have their urlid be mycompany or mc or a38985. 

I certainty agree with you with the consistency. It annoys me. 0/5 if you want me to fix it before merge.

jwilander replied:
I've been going with Url over URL for all my variable names

jwilander replied:
It's kind of the team identifier but I guess we can't call it that, maybe TeamHandle?

crspeller replied:
We are going to go with Name and DisplayName to be consistent with channels. Corey is 5/5 on this.

coreyhulen commented on model/session.go:
```diff
@@ -29,6 +29,7 @@ type Session struct {
 	LastActivityAt int64     `json:"last_activity_at"`
 	UserId         string    `json:"user_id"`
 	TeamId         string    `json:"team_id"`
+	TeamURL        string    `json:"team_url"`
```
Why do I ned this if I have TeamId?  wonder if this should just be Name (see comment below).

jwilander replied:
Yeah and I think changing the TeamUrl in the Context struct to be just TeamName or TeamId would be useful as well because there will be requests that don't have a session but are still doing something on a specific team

coreyhulen commented on web/react/components/delete_channel_modal.jsx:
```diff
@@ -12,7 +12,7 @@ module.exports = React.createClass({
             function(data) {
                 AsyncClient.getChannels(true);
-                window.location.href = '/channels/town-square';
+                window.location.href = '/';
```
should this have /teamName/

crspeller replied:
/ redirects to your team now so it is cleaner.

coreyhulen commented on web/react/components/login.jsx:
```diff
@@ -31,11 +32,11 @@ var FindTeamDomain = React.createClass({
                 console.log(data);
                 if (data) {
-                    window.location.href = window.location.protocol + "//" + domain + "." + utils.getDomainWithOutSub();
+					window.location.href = req.getResponseHeader(Constants.TEAM_URL_HEADER) + '/channels/town-square';
```
Not sure why this needs to be in the header.  Should be in the UserStore?  or some store where we already have the team object.

crspeller replied:
See comment above. But not having it in the store allows someone to access the site from different domains. Such as IP address or domain.

coreyhulen commented on web/react/stores/browser_store.jsx:
```diff
@@ -8,7 +8,7 @@ function getPrefix() {
 
 // Also change model/utils.go ETAG_ROOT_VERSION
-var BROWSER_STORE_VERSION = '.3';
+var BROWSER_STORE_VERSION = '.4';
```
should also change etag version?

crspeller replied:
Yes will fix.

crspeller replied:
Actually this was intentional because I didn't want merge conflicts. I figured we would change it after all these PRs go in. (I thought they might change it as well)

jwilander commented on api/file.go:
```diff
@@ -115,7 +115,7 @@ func uploadFile(c *Context, w http.ResponseWriter, r *http.Request) {
 		}
 
-		fileUrl := c.TeamUrl + "/api/v1/files/get/" + channelId + "/" + c.Session.UserId + "/" + uid + "/" + url.QueryEscape(files[i].Filename)
+		fileUrl := c.SiteUrl + "/api/v1/files/get/" + channelId + "/" + c.Session.UserId + "/" + uid + "/" + url.QueryEscape(files[i].Filename)
```
Does this URL work? Shouldn't it need a /<team>/ somewhere in the URL?

crspeller replied:
The `/api/*` URLs have not changed to add a `/<team>/`  They will always have the required data and are not user facing.

jwilander commented on api/team.go:
```diff
@@ -515,11 +371,11 @@ func InviteMembers(team *model.Team, user *model.User, invites []string) {
 			data := model.MapToJson(props)
 			hash := model.HashPassword(fmt.Sprintf("%v:%v", data, utils.Cfg.ServiceSettings.InviteSalt))
-			bodyPage.Props["Link"] = fmt.Sprintf("%s/signup_user_complete/?d=%s&h=%s", teamUrl, url.QueryEscape(data), url.QueryEscape(hash))
+			bodyPage.Props["Link"] = fmt.Sprintf("%s/signup_user_complete/?d=%s&h=%s", c.SiteUrl, url.QueryEscape(data), url.QueryEscape(hash))
```
Doesn't need a ```/<team>/``` in the url?

jwilander commented on config/config_docker.json:
```diff
@@ -40,10 +39,6 @@
         "S3SecretAccessKey": "",
         "S3Bucket": "",
         "S3Region": "",
```
I think you need to remove the last comma here

crspeller replied:
It works fine, but I will fix.

jwilander commented on web/react/components/password_reset.jsx:
```diff
@@ -119,26 +94,15 @@ ResetPassword = React.createClass({
     },
     render: function() {
-        var update_text = this.state.update_text ? <div className="form-group"><br/><label className="control-label reset-form">{this.state.update_text} Click <a href="/login">here</a> to log in.</label></div> : null;
+		var update_text = this.state.update_text ? <div className="form-group"><br/><label className="control-label reset-form">{this.state.update_text} Click <a href={"/" + this.pros.urlId + "/login"}>here</a> to log in.</label></div> : null;
```
Minor, but weird indenting here?

crspeller replied:
Hmm something is wrong with my vimrc and jsx.

jwilander commented on web/react/components/password_reset.jsx:
```diff
@@ -162,13 +126,14 @@ module.exports = React.createClass({
             return (
                 <SendResetPasswordLink
                     teamName={this.props.teamName}
+					urlId={this.props.urlId}
```
Weird indenting again

jwilander commented on web/react/components/sidebar.jsx:
```diff
@@ -383,7 +297,7 @@ var SidebarLoggedIn = React.createClass({
             } else {
                 return (
-                    <li key={channel.name} className={active}><a className={"sidebar-channel " + titleClass} href={"/channels/"+channel.name}><span className="status" dangerouslySetInnerHTML={{__html: statusIcon}} /> {badge}{channel.display_name}</a></li>
+					<li key={channel.name} className={active}><a className={"sidebar-channel " + titleClass} href={"/" + TeamStore.getCurrentUrlId() + "/channels/"+channel.name}><span className="status" dangerouslySetInnerHTML={{__html: statusIcon}} /> {badge}{channel.display_name}</a></li>
```
More weird indenting

jwilander commented on web/react/components/sidebar_header.jsx:
```diff
@@ -73,19 +73,15 @@ var NavbarDropdown = React.createClass({
+				teams.push(<li key={ urlid }><a href={window.location.origin + "/" + urlid }>Switch to { urlid }</a></li>);
             }
         }
+		teams.push(<li><a href={window.location.origin + "/signup_team" }>Create a New Team</a></li>);
```
Weird indenting again

jwilander commented on web/react/pages/home.jsx:
```diff
@@ -2,13 +2,15 @@
+        window.location = teamURL + "/channels/" + Constants.DEFAULT_CHANNEL;
     } else {
-        window.location.replace("/channels/" + last);
+		console.log("redir to: " +  teamURL);
```
Is this console log left in on purpose?

crspeller replied:
Nope. Will remove.

crspeller commented:
### Done
- Renamed team.Name to DisplayName
- Renamed team.URLId to team.Name
- Fixed inenting
- Rebased

### Todo
- Remove response header and create team store method to get teamURL
- Remove teamURL from session

crspeller commented:
Ready for review. 

#### Changes
- Renamed team.Name to DisplayName
- Renamed team.URLId to team.Name
- Fixed inenting
- Rebased
- Remove response header and create team store method to get teamURL
- Remove teamURL from session
- Rebased again

coreyhulen commented on api/context.go:
```diff
@@ -299,6 +314,39 @@ func (c *Context) SetUnknownError(where string, details string) {
+	c.teamURLValid = valid
+}
+
+func (c *Context) setTeamURLFromSession() {
```
shouldn't be a blocking call.

crspeller replied:
If this does not block, GetTeamURL will print the debug message if it wins the race.

coreyhulen commented on web/web.go:
```diff
@@ -42,25 +42,28 @@ func (me *HtmlTemplatePage) Render(c *api.Context, w http.ResponseWriter) {
+	// Bug in gorilla.mux pervents us from using regex here.
+	mainrouter.Handle("/{team}/channels/{channelname}", api.UserRequired(getChannel)).Methods("GET")
+
+	// Anything added here must have an _ in it so it does not conflict with team names
```
I wonder if these should be moved above the handlers with "{team:" stuff just in case.  I think first route wins?  vs depending on _.  my fear is someone will change the code without understanding the subtly. (0/5)
